pkg/agent/proxy/proxier.go

```diff
@@ -0,0 +1,268 @@
+	// current is state after applying all of those.
+	endpointsChanges *endpointsChangesTracker
+	serviceChanges   *serviceChangesTracker
+	// syncProxyRulesMutex protects internal caches and states
+	syncProxyRulesMutex sync.Mutex
+	// serviceMap stores services we expected to be installed.
+	serviceMap upstream.ServiceMap
+	// serviceInstalledMap stores services we actually installed.
+	serviceInstalledMap upstream.ServiceMap
+	// endpointsMap stores endpoints we expected to be installed.
+	endpointsMap types.EndpointsMap
+	// endpointInstalledMap stores endpoints we actually installed.
+	endpointInstalledMap map[upstream.ServicePortName]map[string]struct{}
+	groupCounter         types.GroupCounter
+
+	runner       *upstream.BoundedFrequencyRunner
+	stopChan     <-chan struct{}
+	agentQuerier querier.AgentQuerier
+	ofClient     openflow.Client
+}
+
+func (p *proxier) isInitialized() bool {
+	return p.endpointsChanges.Synced() && p.serviceChanges.Synced()
+}
+
+func (p *proxier) removeStaleServices() {
+	for svcPortName, svcPort := range p.serviceInstalledMap {
+		if _, ok := p.serviceMap[svcPortName]; ok {
+			continue
+		}
+		svcInfo := svcPort.(*types.ServiceInfo)
+		if err := p.ofClient.UninstallServiceFlows(svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFTransportProtocol); err != nil {
+			klog.Errorf("Failed to remove flows of Service %v: %v", svcPortName, err)
+			continue
+		}
+		endpoints := p.endpointsMap[svcPortName]
+		var endpointsList []upstream.Endpoint
+		for _, endpoint := range endpoints {
+			endpointsList = append(endpointsList, endpoint)
+		}
+		if err := p.ofClient.UninstallEndpointsFlows(svcInfo.OFTransportProtocol, endpointsList...); err != nil {
+			klog.Errorf("Failed to remove flows of Service Endpoints %v: %v", svcPortName, err)
+			continue
+		}
+		groupID, _ := p.groupCounter.Get(svcPortName)
+		if err := p.ofClient.UninstallServiceGroup(groupID); err != nil {
+			klog.Errorf("Failed to remove flows of Service %v: %v", svcPortName, err)
+			continue
+		}
+		delete(p.serviceInstalledMap, svcPortName)
+	}
+}
+
+func (p *proxier) removeStaleEndpoints(staleEndpoints map[upstream.ServicePortName]map[string]upstream.Endpoint) {
+	for svcPortName, endpoints := range staleEndpoints {
+		bindingProtocol := binding.ProtocolTCP
+		if svcPortName.Protocol == corev1.ProtocolUDP {
+			bindingProtocol = binding.ProtocolUDP
+		} else if svcPortName.Protocol == corev1.ProtocolSCTP {
+			bindingProtocol = binding.ProtocolSCTP
+		}
+		for _, endpoint := range endpoints {
+			if err := p.ofClient.UninstallEndpointsFlows(bindingProtocol, endpoint); err != nil {
+				klog.Errorf("Error when removing Endpoint %v for %v", endpoint, svcPortName)
+				continue
+			}
+			if m, ok := p.endpointInstalledMap[svcPortName]; ok {
+				delete(m, endpoint.String())
+				if len(m) == 0 {
+					delete(p.endpointInstalledMap, svcPortName)
+				}
+			}
+		}
+	}
+}
+
+func (p *proxier) localPodsL3L2Mapping() map[string]net.HardwareAddr {
```

Comment:
I wonder why we need to rewrite DMAC specially for service access traffic when we already have `L3Forwarding` rewrite MAC, though the table name doesn't fit.
But in theory we can just DNAT in ServiceLB table and leave the following table to rewrite DMAC based on DestIP?

Reply:
Discussed with @wenyingd offline, we think it's doable and can make proxy get rid of InterfaceStore and the IP to MAC query logic, and you can also call `InstallEndpointFlows` in batch.
If you want to distinguish local endpoints and remote endpoints, you can use `Endpoint.GetIsLocal`.
To make L3forwardingTable rewrite MAC for service traffic, we discussed two options:
1. Rewrite DMAC to global VMAC in ServiceLB table so that L3forwarding will rewrite DMAC
2. Use a reg bit to indicate whether this packet needs MAC rewrite, and set it for service traffic and tunnel traffic, and change the condition of MAC rewrite in L3forwardingTable.

The option 1 might be riskless and good as a short term solution. I would suggest using option 2 in long term as it's more readable, can be used for other scenarios that need MAC rewrite, and perhaps more performant comparing to rewriting the whole MAC which is 6 bytes.

Reply:
Thanks for your informative description. I would prefer option 1 since we are approaching the release date and it cost less to be verified. For option 2, I think we can enhance it in future.
I have just implemented the option 1, PTAL.